File: sentil-cpp/tests/test_formula.cpp
#include <sentil/sentil.hpp>

#include <string>

#include "sentil_test.hpp"

using sentil::Expr;
using sentil::Formula;
using sentil::ProbabilityOp;
using sentil::Trace;

static std::string json_of(const Formula& f) { return f.to_json(); }

int main() {
    Formula phi = Formula::parse("always[0,10](x > 0)");
    CHECK(phi.is_temporal());
    CHECK(phi.depth() >= 2);
    CHECK(phi.variables().size() == 1 && phi.variables()[0] == "x");
    CHECK(Formula::from_json(phi.to_json()).to_json() == phi.to_json());

    CHECK(json_of((Expr::var("x") * 2.0) > 5.0) == json_of(Formula::parse("x * 2 > 5")));
    CHECK(json_of(5.0 < Expr::var("x")) == json_of(Formula::parse("5 < x")));
    CHECK(json_of(sentil::abs(Expr::var("y")) <= 3.0) == json_of(Formula::parse("abs(y) <= 3")));
    CHECK(json_of((Expr::var("a") - Expr::var("b")) != 0.0) ==
          json_of(Formula::parse("a - b != 0")));

    CHECK(json_of(!(Expr::var("x") > 0.0)) == json_of(Formula::parse("not(x > 0)")));
    CHECK(json_of((Expr::var("x") > 0.0) && (Expr::var("y") > 0.0)) ==
          json_of(Formula::parse("(x > 0) and (y > 0)")));
    CHECK(json_of((Expr::var("x") > 0.0) || (Expr::var("y") > 0.0)) ==
          json_of(Formula::parse("(x > 0) or (y > 0)")));
    CHECK(json_of(sentil::implies(Expr::var("x") > 0.0, Expr::var("y") > 0.0)) ==
          json_of(Formula::parse("(x > 0) implies (y > 0)")));
    CHECK(json_of(sentil::next(Expr::var("x") > 0.0)) == json_of(Formula::parse("next(x > 0)")));
    CHECK(json_of(sentil::always(Expr::var("x") > 0.0, 0, 10)) ==
          json_of(Formula::parse("always[0,10](x > 0)")));
    CHECK(json_of(sentil::eventually(Expr::var("x") > 0.0)) ==
          json_of(Formula::parse("eventually(x > 0)")));
    CHECK(json_of(sentil::until(Expr::var("x") > 0.0, Expr::var("y") > 0.0, 0, 2)) ==
          json_of(Formula::parse("(x > 0) until[0,2] (y > 0)")));
    CHECK(json_of((Expr::var("x") > 0.0).always(0, 10)) ==
          json_of(Formula::parse("always[0,10](x > 0)")));
    CHECK(json_of(sentil::probability(sentil::always(Expr::var("x") > 0.0), ProbabilityOp::Ge,
                                      0.95)) == json_of(Formula::parse("P>=0.95(always(x > 0))")));

    Trace t = Trace::indexed(3);
    t.add_signal("x", {3, 1, 2});
    Formula bounded = Formula::parse("always[0,2](x > 0)");
    CHECK_BITS(bounded.robustness(t), 1.0);
    std::vector<double> sig = bounded.robustness_signal(t);
    CHECK(sig.size() == 3);
    CHECK_BITS(sig[0], 1.0);
    CHECK_BITS(sig[2], 2.0);

    Trace dip = Trace::indexed(3);
    dip.add_signal("x", {1, -2, 3});
    CHECK(!Formula::parse("always(x > 0)").violations(dip).empty());

    return sentil_report("test_formula");
}

File: sentil-jl/test/test_formula.jl
using Test
using Sentil

@testset "formula building" begin
    x = variable("x")
    y = variable("y")
    pairs = [
        (variable("x") + 2 > 5, "x + 2 > 5"),
        (3 - variable("x") >= 0, "3 - x >= 0"),
        (variable("x") * variable("y") <= 10, "x * y <= 10"),
        (variable("x")^2 == 4, "x ^ 2 == 4"),
        (abs(variable("x")) < 1, "abs(x) < 1"),
        (min(variable("x"), variable("y")) != 0, "min(x, y) != 0"),
        (sqrt(variable("x")) > 2, "sqrt(x) > 2"),
        (!(variable("x") > 0), "not (x > 0)"),
        ((variable("x") > 0) & (variable("y") < 1), "(x > 0) and (y < 1)"),
        (implies(variable("x") > 0, variable("y") < 1), "(x > 0) -> (y < 1)"),
        (always(variable("x") > 0; lower = 0, upper = 10), "always[0, 10](x > 0)"),
        (eventually(variable("x") > 5; upper = 3), "eventually[0, 3](x > 5)"),
        (until(variable("x") > 0, variable("y") > 0; upper = 5), "(x > 0) until[0, 5] (y > 0)"),
        (probability(always(variable("x") > 0), ProbabilityOp.Ge, 0.9), "P>=0.9 (always (x > 0))"),
    ]
    for (built, text) in pairs
        @test to_json(built) == to_json(formula(text))
    end
end

@testset "formula introspection" begin
    f = formula("always[0, 10](x > 5 and y < 3)")
    @test depth(f) > 1
    @test is_temporal(f)
    @test !is_temporal(formula("x > 0"))
    @test variables(f) == ["x", "y"]
    @test to_json(from_json(Formula, to_json(f))) == to_json(f)
end

@testset "expression math vs parser" begin
    t = Trace([0.0, 1.0, 2.0], "x", [4.0, 1.0, 9.0])
    for (built, text) in [(sqrt(variable("x")) > 1, "sqrt(x) > 1"),
                          (variable("x") % 4 < 2, "x % 4 < 2"),
                          (pow(variable("x"), 2) > 5, "x ^ 2 > 5")]
        @test robustness_signal(built, t) == robustness_signal(formula(text), t)
    end
end